### PR TITLE
Safe lastOption for java concurrent map

### LINKED ISSUE #12572: NoSuchElementException throw from lastOption
## Reproduction steps

Scala version: 2.13.x

```scala
  def main(args: Array[String]): Unit = {


    val map: ConcurrentSkipListMap[Int, Int] = new ConcurrentSkipListMap[Int, Int]();
    map.put(1, 1);
    map.put(2, 3);
    map.put(3, 4);
    val subMap = map.subMap(1, 3)
    val values = subMap.values().asScala
    
    // Thead-A 
    values.lastOption.get

    // Thead-B 
    map.clear()
  }

```
it will occur under specific circumstances....... (Not necessarily)

## Problem
when i use concurrent data structure, Iterable#lastOption maybe throw NoSuchElementException

image

According to the note of method, it did not  declare will throw NoSuchElementException.  i think we should avoid that

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12572

### CODE REVIEW

**Comment by som-snytt** [comment]
`headOption` doesn't use `isEmpty`, but uses `iterator`, which guarantees an element if `hasNext`.

Quick review suggests maybe there are no other methods that are sneakily non-atomic. But there are a lot of collection methods.

**Comment by lrytz** [comment]
`TrieMap.lastOption` is also inherited from `IterableOps`, does it have the same issue? Maybe we can find other operations with similar issues?

File: src/library/scala/collection/convert/JavaCollectionWrappers.scala [hunk 1]
```diff
@@ -442,8 +444,13 @@ private[collection] object JavaCollectionWrappers extends Serializable {
...
     def replace(k: K, v: V): Option[V] = Option(underlying.replace(k, v))
 
-    def replace(k: K, oldvalue: V, newvalue: V): Boolean =
-      underlying.replace(k, oldvalue, newvalue)
+    def replace(k: K, oldvalue: V, newvalue: V): Boolean = underlying.replace(k, oldvalue, newvalue)
+
+    override def lastOption: Option[(K, V)] =
+      underlying match {
+        case nav: NavigableMap[K @unchecked, V @unchecked] => Option(nav.lastEntry).map(e => (e.getKey, e.getValue))
+        case _ => Try(super.lastOption).toOption.flatten
```
Comment (lrytz) [inline src/library/scala/collection/convert/JavaCollectionWrappers.scala]: could this go as an override to `collection.concurrent.Map`?

Comment (som-snytt) [reply src/library/scala/collection/convert/JavaCollectionWrappers.scala]: I will look into it.

**Comment by som-snytt** [comment]
Obvious in retrospect,
```
[error] - library/mimaReportBinaryIssues
[error]   - library/mimaReportBinaryIssues failed: java.lang.RuntimeException: Failed binary compatibility check against org.scala-lang:scala-library:2.13.8! Found 1 potential problems (filtered 5)
[error] Saved log of library/mimaReportBinaryIssues:
[error] [error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.13.8! Found 1 potential problems (filtered 5)
[error] [error]  * in other version, classes mixing scala.collection.concurrent.Map need be recompiled to wire to the new static mixin forwarder method all super calls to method lastOption()scala.Option
[error] [error]    filter with: ProblemFilters.exclude[NewMixinForwarderProblem]("scala.collection.concurrent.Map.lastOption")
[error] Failure due to previous errors
[error] (testAll) Failure due to previous errors
```
And I'd already squashed...

**Comment by som-snytt** [comment]
Compare https://github.com/scala/scala-parallel-collections/blob/main/core/src/main/scala/scala/collection/parallel/ParIterableLike.scala#L223 which also uses the brittle idiom for `headOption`.

**Review by lrytz — APPROVED** [review]
Thank you!

**Pull request merged**